tactics/autorewrite.ml

```diff
@@ -278,5 +285,23 @@ let add_rew_rules ~locality base lrul =
   let open Goptions in
   match locality with
   | OptLocal -> cache_hintrewrite ((),(base,lrul))
-  | OptDefault | OptGlobal -> Lib.add_anonymous_leaf (inGlobalHintRewrite (base,lrul))
-  | OptExport -> Lib.add_anonymous_leaf (inExportHintRewrite (base,lrul))
+  | OptDefault ->
+    let () =
+      if not @@ Global.sections_are_opened () then
+        warn_deprecated_hint_rewrite_without_locality ()
+    in
+    Lib.add_anonymous_leaf (inGlobalHintRewrite (base,lrul))
+  | OptGlobal ->
+    let () =
+      if Global.sections_are_opened () then
+      CErrors.user_err Pp.(str
+        "This command does not support the global attribute in sections.");
+    in
+    Lib.add_anonymous_leaf (inGlobalHintRewrite (base,lrul))
+  | OptExport ->
+    let () =
+      if Global.sections_are_opened () then
+        CErrors.user_err Pp.(str
+          "This command does not support the export attribute in sections.");
+    in
+    Lib.add_anonymous_leaf (inExportHintRewrite (base,lrul))
```

Comment:
I would also change this to (purely for readability):
```suggestion
  | OptGlobal when Global.sections_are_opened () ->
    CErrors.user_err Pp.(str
      "This command does not support the global attribute in sections.");
  | OptGlobal ->
    Lib.add_anonymous_leaf (inGlobalHintRewrite (base,lrul))
  | OptExport when Global.sections_are_opened () ->
      CErrors.user_err Pp.(str
        "This command does not support the export attribute in sections.");
  | OptExport ->
    Lib.add_anonymous_leaf (inExportHintRewrite (base,lrul))
```

Reply:
I really hate `when` clauses, they are a great way to make the code fragile. The gain in legibility is quite low in this case and probably more error-prone than anything else.

Reply:
I feel like you think they are a way to make the code fragile in general, but this is a clear case where it would make it *less* fragile. Anyway, this is not a blocker for this PR, so I'm OK to merge as is but I encourage you strongly to refactor the whole thing in the future.

Reply:
I'll replace the object duplication at some point, but not in this PR.